File: ruby/specs/login_specs/login_spec.rb
require_relative '../../spec_helper.rb'

describe 'user login', type: :feature do
  include LogInHelper

  before(:each) do
    @pages = Pages.new
    load_credentials
  end

  let(:login_page){ @pages.login_page }

  it 'logs the user in' do
    login_page.load

    login_page.user_name.visible?
    login_page.user_name.send_keys @user_name
    login_page.password.visible?
    login_page.password.send_keys @password
    login_page.submit_button.visible?
    login_page.submit_button.click

    expect(@pages.logout_page.log_out_button.visible?)
  end

  it 'handles incorrect user_name' do
    login_page.load

    login_page.user_name.visible?
    login_page.user_name.send_keys "incorrectUser"
    login_page.password.visible?
    login_page.password.send_keys @password
    login_page.submit_button.visible?
    login_page.submit_button.click

    expect(login_page.error_message.text).to eq('Your username is invalid!')
  end

  it 'handles incorrect password' do
    login_page.load

    login_page.user_name.visible?
    login_page.user_name.send_keys @user_name
    login_page.password.visible?
    login_page.password.send_keys 'incorrectPassword'
    login_page.submit_button.visible?
    login_page.submit_button.click

    expect(login_page.error_message.text).to eq('Your password is invalid!')
  end

  it 'handles blank form submission' do
    login_page.load
    login_page.submit_button.click

    expect(login_page.error_message.text).to eq('Your username is invalid!')
  end

end

File: playwright_js/specs/login.spec.js
import { test as baseTest, expect } from '@playwright/test';
import { LogInHelper } from '../helpers/log_in_helper.js';
import { LoginPage } from '../page_objects/login_page.js';
import { LogoutPage } from '../page_objects/logout_page.js';

// Fixture to handle loading credentials and load page objects
const test = baseTest.extend({
    credentials: async ( { }, use) => {
        const login_credentials = new LogInHelper();
        await login_credentials.init();
        await use(login_credentials);
    },
    loginPage: async ( { page }, use) => {
        const login_page = new LoginPage(page);
        await use(login_page);
    },
    logoutPage: async ( { page }, use ) => {
        const logout_page = new LogoutPage(page);
        await use(logout_page);
    }
});

test.describe('Login Page', () => {

    test('logs the user in', async ({ credentials, loginPage, logoutPage }) => { 
        await loginPage.goto();

        await loginPage.user_name_field().fill(credentials.user_name());
        await loginPage.password_field().fill(credentials.password());
        await loginPage.submit_button().click();

        await expect(logoutPage.log_out_button()).toBeVisible();
    });

    test('handles incorrect user_name', async ({ credentials, loginPage }) => { 
        await loginPage.goto();

        await loginPage.user_name_field().fill("incorrectUser");
        await loginPage.password_field().fill(credentials.password());
        await loginPage.submit_button().click();

        await expect(loginPage.error_message()).toHaveText('Your username is invalid!')
    });

    test('handles incorrect password', async ({ credentials, loginPage }) => { 
        await loginPage.goto();

        await loginPage.user_name_field().fill(credentials.user_name());
        await loginPage.password_field().fill('incorrectPassword');
        await loginPage.submit_button().click();

        await expect(loginPage.error_message()).toHaveText('Your password is invalid!')
    });

    test('handles blank form submission', async ({ loginPage }) => { 
        await loginPage.goto();

        await loginPage.submit_button().click();
        await expect(loginPage.error_message()).toHaveText('Your username is invalid!')
    });
});